File: java/DeleteCode.java
// Delete Code > Create Code
// Prefer to delete code rather than create code.
// Sometimes even new features can be created by deleting code.

public class DeleteCode {
    // Before
    static class NotificationServiceBefore {
        public void sendNotification(UserBefore user, String message) {
            if (user.emailNotificationsEnabled) {
                sendEmail(user.email, message);
            } else {
                System.out.println("Email notifications are disabled for " + user.name + ".");
            }

            if (user.smsNotificationsEnabled) {
                sendSms(user.phoneNumber, message);
            } else {
                System.out.println("SMS notifications are disabled for " + user.name + ".");
            }
        }

        private void sendEmail(String email, String message) {
            // Code to send email
            System.out.println("Email sent to " + email + ": " + message);
        }

        private void sendSms(String phoneNumber, String message) {
            // Code to send SMS
            System.out.println("SMS sent to " + phoneNumber + ": " + message);
        }
    }

    static class UserBefore {
        public String name;
        public String email;
        public String phoneNumber;
        public boolean emailNotificationsEnabled = false;
        public boolean smsNotificationsEnabled = false;

        public UserBefore(String name, String email, String phoneNumber) {
            this.name = name;
            this.email = email;
            this.phoneNumber = phoneNumber;
        }
    }

    // After
    static class NotificationService {
        public void sendNotification(User user, String message) {
            sendEmail(user.email, message);
            sendSms(user.phoneNumber, message);
        }

        private void sendEmail(String email, String message) {
            // Code to send email
            System.out.println("Email sent to " + email + ": " + message);
        }

        private void sendSms(String phoneNumber, String message) {
            // Code to send SMS
            System.out.println("SMS sent to " + phoneNumber + ": " + message);
        }
    }

    static class User {
        public String name;
        public String email;
        public String phoneNumber;

        public User(String name, String email, String phoneNumber) {
            this.name = name;
            this.email = email;
            this.phoneNumber = phoneNumber;
        }
    }

    public static void main(String[] args) {
        // Before usage
        UserBefore userBefore = new UserBefore("Alice", "alice@example.com", "123-456-7890");
        userBefore.emailNotificationsEnabled = true;
        userBefore.smsNotificationsEnabled = true;
        NotificationServiceBefore serviceBefore = new NotificationServiceBefore();
        serviceBefore.sendNotification(userBefore, "Your order has been shipped.");

        System.out.println();

        // After usage
        User user = new User("Alice", "alice@example.com", "123-456-7890");
        NotificationService service = new NotificationService();
        service.sendNotification(user, "Your order has been shipped.");
    }
}

File: rust/delete_code.rs
// Delete Code > Create Code
// Prefer to delete code rather than create code.
// Sometimes even new features can be created by deleting code.

// Before
struct NotificationServiceBefore;

impl NotificationServiceBefore {
    fn send_notification(&self, user: &UserBefore, message: &str) {
        if user.email_notifications_enabled {
            self.send_email(&user.email, message);
        } else {
            println!("Email notifications are disabled for {}.", user.name);
        }

        if user.sms_notifications_enabled {
            self.send_sms(&user.phone_number, message);
        } else {
            println!("SMS notifications are disabled for {}.", user.name);
        }
    }

    fn send_email(&self, email: &str, message: &str) {
        // Code to send email
        println!("Email sent to {}: {}", email, message);
    }

    fn send_sms(&self, phone_number: &str, message: &str) {
        // Code to send SMS
        println!("SMS sent to {}: {}", phone_number, message);
    }
}

struct UserBefore {
    name: String,
    email: String,
    phone_number: String,
    email_notifications_enabled: bool,
    sms_notifications_enabled: bool,
}

impl UserBefore {
    fn new(name: String, email: String, phone_number: String) -> Self {
        UserBefore {
            name,
            email,
            phone_number,
            email_notifications_enabled: false,
            sms_notifications_enabled: false,
        }
    }
}

// After
struct NotificationService;

impl NotificationService {
    fn send_notification(&self, user: &User, message: &str) {
        self.send_email(&user.email, message);
        self.send_sms(&user.phone_number, message);
    }

    fn send_email(&self, email: &str, message: &str) {
        // Code to send email
        println!("Email sent to {}: {}", email, message);
    }

    fn send_sms(&self, phone_number: &str, message: &str) {
        // Code to send SMS
        println!("SMS sent to {}: {}", phone_number, message);
    }
}

struct User {
    name: String,
    email: String,
    phone_number: String,
}

impl User {
    fn new(name: String, email: String, phone_number: String) -> Self {
        User {
            name,
            email,
            phone_number,
        }
    }
}

fn main() {
    // Before usage
    let mut user_before = UserBefore::new(
        "Alice".to_string(),
        "alice@example.com".to_string(),
        "123-456-7890".to_string(),
    );
    user_before.email_notifications_enabled = true;
    user_before.sms_notifications_enabled = true;
    let service_before = NotificationServiceBefore;
    service_before.send_notification(&user_before, "Your order has been shipped.");

    println!();

    // After usage
    let user = User::new(
        "Alice".to_string(),
        "alice@example.com".to_string(),
        "123-456-7890".to_string(),
    );
    let service = NotificationService;
    service.send_notification(&user, "Your order has been shipped.");
}
